ios/viewmodels/view-model-reuse.md

```diff
@@ -0,0 +1,167 @@
+
+
+### Protocol based view model
+
+We can use protocols to standardize the interface for the view, while keeping implementation details abstracted out into it's own entity.
+
+```swift
+
+protocol TitleMessageCellConforming {
+  let title: String { get }
+  let message: String { get }
+  let showsEditButton: Bool { get }
+}
+
+```
+
+Note that the only difference between the protocol and the concrete implementation of this class is the type. With this, the details of each cell in ReviewViewController can be abstracted out.
+
+```swift
+
+// PhoneNumberTitleMessageCellViewModel.swift
+struct PhoneNumberTitleMessageCellViewModel: TitleMessageCellConforming {
+  let title: String
+  let message: String
+  let showsEditButton: Bool
+
+  init(profile: Profile) {
+    let formatter = PhoneNumberFormatter(forCountryCode: currentCountryCode)
+    let number = formatter.formatNumber(profile.phoneNumber) ?? "-"
+    title = "Contact number"
+    message = number
+    showsEditButton = profile.canEdit
+  }
+
+// TopicsTitleMessageCellViewModel.swift
+struct TopicsTitleMessageCellViewModel: TitleMessageCellConforming {
+  let title: String
+  let message: String
+  let showsEditButton: Bool
+
+  init(topics: [Topics], review: Review) {
+    title = "Visit Details"
+    message = topics.filter({ $0.selected }).map({ $0.title }).joinWithSeparator("\n")
+    showsEditButton = review.canEdit
+  }
+}
+
+```
+
+This moves view logic code into the view models, leaving a very nice and clean implementation for our view controller:
+
+```swift
+
+// ReviewViewController.swift
+func viewModel(forIndex index: Int) {
+  switch index {
+    case 0:
+      return PhoneNumberTitleMessageCellViewModel(profile)
+    case 1:
+      return AllergiesTitleMessageCellViewModel(allergies, review: review)
+
+    ...
+  }
+}
+
+```
+
+For the profile, it might still be okay to leave the view model logic in the profile view controller, due to it's simplicity.
+
+```swift
+
+// ProfileViewController.swift
+func viewModel(forIndex index: Int) {
+  switch index {
+    case 0:
+      return TitleMessageCellViewModel(title: "First Name", message: profile.firstName) // Now, TitleMessageCellViewModel is simply a struct that implements TitleMessageCellConforming.
+    case 1:
+      return TitleMessageCellViewModel(title: "Surname", message: profile.lastName)
+    ...
+  }
+}
+
+```
+
+## Conclusions and takeaways
+
+View models are a tool to facilitate view logic abstraction, but also code reuse. A protocol-based approach allow us to take abstraction one level further, so that our view relies on a generic interface, without forcing an implementation.
+
+In the protocol based implementation, we didn't remove the logic completely from the view controller, although we could have just have easily created view models for each type of 'row' in ProfileViewController. While this leaves view logic in the view controller, we have largely succeeded in making TitleMessageCell reusable.
```

Comment:
As a learning tool, I agree that maybe we should document 'the best of the best' but I do think that the danger of making blanket statements like every 'view should have a view model'(if i'm understnding you correctly) doesn't capture the spirit of what we are striving for - to reduce bloat, and to encourage separation of concerns. There are other ways of achieving this - view models are just one of them.

I think with any architecture, we can take it to space - we need to encourage balance and thoughtful use. Maybe this particular examples don't do that - and they are definitely up for change.